pkg/logql/log/metrics_extraction.go

```diff
@@ -226,3 +230,75 @@ func convertBytes(v string) (float64, error) {
 	}
 	return float64(b), nil
 }
+
+// ParserHint are hints given to LogQL parsers.
+// This is specially useful for parser that extract implicitly all possible label keys.
+// This is used only within metric queries since it's rare that you need all label keys.
+// For example in the following expression:
+//
+//		sum by (status_code) (rate({app="foo"} | json [5m]))
+//
+// All we need to extract is the status_code in the json parser.
+type ParserHint interface {
+	// Tells is a label with the given key should be extracted.
+	ShouldExtract(key string) bool
+	// Tells if there's any hint that start with the given prefix.
+	// This allows to speed up key searching in nested structured like json.
+	ShouldExtractPrefix(prefix string) bool
+	// Tells if we should not extract any labels.
+	// For example in :
+	//		 sum(rate({app="foo"} | json [5m]))
+	// We don't need to extract any labels from the log line.
+	NoLabels() bool
+}
+
+type parserHint struct {
+	noLabels       bool
+	requiredLabels []string
+}
+
+func (p *parserHint) ShouldExtract(key string) bool {
+	if len(p.requiredLabels) == 0 {
+		return true
+	}
+	for _, l := range p.requiredLabels {
+		if l == key {
+			return true
+		}
+	}
+	return false
+}
+
+func (p *parserHint) ShouldExtractPrefix(prefix string) bool {
+	if len(p.requiredLabels) == 0 {
+		return true
+	}
+	for _, l := range p.requiredLabels {
+		if strings.HasPrefix(l, prefix) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (p *parserHint) NoLabels() bool {
+	return p.noLabels
+}
+
+// newParserHint creates a new parser hint using the list of labels that are seen and required in a query.
+func newParserHint(requiredLabelNames []string, without, noLabels bool, metricLabelName string) *parserHint {
+	if noLabels {
+		// if we need to extract metric from a label, then use that as required labels list.
+		if metricLabelName != "" {
+			return &parserHint{requiredLabels: []string{metricLabelName}}
```

Comment:
Q: wondering why we need `metricLabelName` as separate param? whoever using parserHint already knows wether its metricLabel or not correct? and they can pass it directly as `requiredLabelNames` in that case? No?

I'm thinking why not something like this?

```
if noLabels {
     return &parserHint{noLabels: true}
}

if without {
		return noParserHints
}

return &parserHint{requiredLabels: requiredLabelNames}
```

Basically I'm thinking is the combination of `noLabels` and only `metricLabelName` is needed? at least it sounds contradicting with the naming.

Reply:
Consider
```
sum(
  sum_over_time(
  {job="loki-prod/query-frontend"}
    |= "metrics.go:83"
    | logfmt
    | unwrap duration(duration)
    [10m]
  )
)
```

In this case, we're sum'ing across all series, looking for an aggregated result with the labels `{}`. However, we _need_ to extract the `duration` field used in the extract, which is the `metricLabelName` here.